core/contractsdk/cpp/example/xuper_relayer.cc

```diff
@@ -0,0 +1,477 @@
+    bool succ = blockHeader->ParseFromString(blockHeaderStr);
+    if (!succ) {
+        ctx->error("parse from string error");
+        return;
+    }
+    std::string blockidBuf = blockHeader->blockid();
+    std::string visualBlockid = std::string(64, 'o');
+    if (!encode(blockidBuf, visualBlockid)) {
+        ctx->error("encode blockid failed");
+        return;
+    }
+    const std::string blockHeaderKey = std::string(blockHeaderBucket) + visualBlockid;
+    std::string tmp;
+    if (ctx->get_object(blockHeaderKey, &tmp)) {
+        ctx->error(visualBlockid + " has existed already");
+        return;
+    }
+    // 判断区块类型
+    std::string preHashBuf = blockHeader->pre_hash();
+    // root链
+    if (preHashBuf.size() == 0) {
+        ctx->error("preHash shouldn't be empty");
+        return;
+    }
+    std::string visualPreHash = std::string(64, 'o');
+    if (!encode(preHashBuf, visualPreHash)) {
+        ctx->error("encode preHash failed");
+        return;
+    }
+    // 判断preBlockcHeader是否存在
+    const std::string preBlockHeaderKey = std::string(blockHeaderBucket) + visualPreHash;
+    std::string preBlockHeaderStr;
+    relayer::InternalBlock* preBlockHeader = new(relayer::InternalBlock);
+    if (!ctx->get_object(preBlockHeaderKey, &preBlockHeaderStr)) {
+        ctx->error("missing preHash:" + visualPreHash);
+        return;
+    }
+    // 在主干上添加
+    if (meta->tip_blockid() == visualPreHash) {
+        blockHeader->set_in_trunk(true);
+        preBlockHeader->set_next_hash(blockHeader->blockid());
+        meta->set_tip_blockid(visualBlockid);
+        meta->set_trunk_height(meta->trunk_height()+1);
+    } else {
+        // 在分支上添加
+        if (preBlockHeader->height()+1 > meta->trunk_height()) {
+            // 分支变主干
+            meta->set_trunk_height(preBlockHeader->height()+1);
+            meta->set_tip_blockid(visualBlockid);
+            blockHeader->set_in_trunk(true);
+            // 处理分叉
+            bool succ = handleFork(ctx, meta->tip_blockid(), visualPreHash, blockHeader->blockid());
+            if (!succ) {
+                ctx->error("handle fork failed");
+                return;
+            }
+        }
+    }
+    // 判断blockid是否正确
+    // 判断矿工签名是否正确
+    // 判断2/3签名是否正确
+    
+    // 更新区块头信息
+    blockHeader->SerializeToString(&blockHeaderStr);
+    if (!ctx->put_object(blockHeaderKey, blockHeaderStr)) {
+        ctx->error("put " + visualBlockid + " failed");
+        return;
+    }
+    // 更新ledger meta信息
+    meta->SerializeToString(&metaStr);
+    if (!ctx->put_object(metaKey, metaStr)) {
+        ctx->error("put ledger meta failed");
+        return;
+    }
+}
+
+DEFINE_METHOD(XuperRelayer, verifyTx) {
+    xchain::Context* ctx = self.context();
+    const std::string blockid = ctx->arg("blockid");
+    const std::string txid = ctx->arg("txid");
+    const std::string proofPathStr = ctx->arg("proofPath");
+    const int txIndex = atoi(ctx->arg("txIndex").c_str());
```

Comment:
We do not verify the raw transaction data?

Reply:
Issue about verify will be added later.